### PR TITLE
Implement base stats

### PR DESCRIPTION
**Pull request by johngro**
So here is my first crack at trying to make the base statistics for a mob available for display next to their deck.  It seems to lay out reasonably well when on a vertical tablet (My group plays with a tablet on a music stand mount next to the table).  I'm not sure it is all that great layout-wise in a normal browser; suggestions and advice are welcome.

This should address issue #1

### CODE REVIEW

File: MainUI.jsx [hunk 1]
```diff
@@ -60,6 +87,10 @@ class MainUI extends React.Component {
     this.props.deckVisible.mutate(mutation);
   }
 
+  handleBaseStatsToggle = () => {
+    this.setState({ showBaseStats: !this.state.showBaseStats });
```
Comment (teisenbe) [inline MainUI.jsx]: I would handle this via props like deckVisibility above (if it's not clear from the code, mutation is a mapping from current state to new state).  This will make it get synchronized through React.   The main change is to add it to `propTypes` above and modify `withStorage` below to define deserialization/default values.

**Review by teisenbe — COMMENTED** [review]
This generally looks good to me, but I think some of the CSS needs to be hammered on a bit more; the attribute column seems to cause some weird alignment stuff when the window is wide enough for multiple columns.  Take a look at Scenario 2 for an example.  If you don't feel like hammering on the CSS, I can take a stab at it

File: AbilityDeck.jsx [hunk 2]
```diff
@@ -12,6 +12,22 @@ import { MONSTER_STATS } from './monster_stats';
 
 import * as css from './style/Card.scss';
 
+function StatIcon(props) {
+  const parse = /^%?([^%]+)%?$/;
+  const match = parse.exec(props.name);
+  const imgSource = `images/${match[1]}.svg`;
```
Comment (teisenbe) [inline AbilityDeck.jsx]: I think we need a special case here for the "flying" attribute, since the image is named fly.svg.  Alternatively, I would be supportive of just renaming it to flying.svg, since it's just referenced in one other place (macros.js).

File: AbilityDeck.jsx [hunk 3]
```diff
@@ -86,22 +103,105 @@ export default function AbilityDeck(props) {
...
+    }
+
+    const statLine = (iconName, data) => {
+      const bossData = (data.length === 1);
+      const empty = (data[0] === 0) && (bossData || (data[1] === 0));
+      if (empty) {
+        return null;
+      }
+
+      const value = bossData ? String(data[0]) : `${data[0]}/${data[1]}`;
```
Comment (teisenbe) [inline AbilityDeck.jsx]: For 0 values, it might look nicer to render them as "-" instead? Also, was chatting with adrake and he suggested maybe handling the non-stat attributes as lists of the form `N: ,,  2` and `E: ,,  2`.  What do you think?

File: MainUI.jsx (hunk 1 above)
Comment (johngro) [reply MainUI.jsx]: sure will do.

I had thought about making it a prop, but I was concerned that this was the type of state that we didn't actually want to serialize.  My worry is that a user would turn on the stats locally, and that suddenly all users  (via the magic of firebase) would have this choice about how a specific user wanted their UI to look reflected onto their layout.  I'm not 100% what is up with the property storage, however, so I might be misreading it.

File: AbilityDeck.jsx (hunk 2 above)
Comment (johngro) [reply AbilityDeck.jsx]: fwiw - I had considered bumping StatIcon into its own file/component and then adding a rather large LUT which would would allow me to remap stat/abilities to the proper icon filename, and then also to add some rich alt text which could do things like remind the user of the specific ability rule along with a page number reference when deep rules lawyer-ing was needed.

I may still end up doing this, but for now, this seems like it might be out of scope for this CL.  I think I'll end up adding a special case for now; both because I might do the LUT anyway someday, and because I would rather not diverge too much from the upstream.  macros.js right now are leftovers from the orignial, so I'd like to avoid introducing diffs whenever I can.

File: AbilityDeck.jsx (hunk 3 above)
Comment (johngro) [reply AbilityDeck.jsx]: re: - instead of 0, sure, np.
re: boolean stats; yes I had considered doing this originally.  What I really wanted to do was to show the icons in either black for normals, and gold for elites.  This sent me down a long path trying to figure out why it was so difficult to do something like a chromakey filter on an SVG (to change the white in the icon to gold).  I eventually decided that "it's complicated" and that I should put something like that off to a future CL.

I also considered splitting the elite and the normal stats, putting one on the left side of the deck and the other on the right.  This way, I could just list all of the numeric stats and then have one line at the bottom for the bool stats.  In the end I though that this would just take too much room, and I would still want to have solved the color coding problem before trying it.

Another option I toyed with was rendering bool stats as "no/yes" instead of "0/1".  In the end, I decided to try to avoid feature creep and just start with this, assuming that it could be changed moving forward.

I'll rework this for now so that we can show them as two sets of bool stats; one for normals and one for elites.  I think it would probably look better with color coding, but I'd like to leave that as a future project for now.

**Comment by johngro** [comment]
hmm..  So have all of the changes in place and working, but I think I might still have some more to do.  There are still more special cases which need to be sorted out.  For example...

- There are stat icons which show up as boolean stats (eg curse) which should not be inverted, even when placed on a white background.  Forest imps in lv 5. version of scenario 7 show this.  Inversion not only flips black/white (good), but also messes up the colors for colored icons (bad).  I think I need to either flesh out that LUT idea to get proper special handling, or look into solving the replace-a-color-in-an-svg idea.
- There are mobs with boolean abilities which cannot be represented using an icon.  For example; Night Demons have "Attackers gain Disadvantage".  These things are not going to fit in the sidebar format I had been going for.  They probably need to be tacked on in the section on the bottom I had called "boss extras", but I need to spend some time thinking about how to add this stuff in a way which does not take too much space.
- There are attributes which are actually composed from multiple attributes.  For example lv 2 elite Flame Demons have "retaliate 2 at range 2".  AFAICT - retaliate+range is the only example of this pattern currently in the DB, so it is unclear how much of a generic solution is really needed here.

TL;DR?  There is more work to be done to make this "perfect".
How do you want me to proceed?  I could upload the WiP for you to look at/play with if you are interested.  If you want, we can move forward with this as is and address its problems in subsequent CLs, or if you would prefer, I can continue to bang on this until I sort out all of the issues I have listed here.

LMK what you want to do.

File: MainUI.jsx (hunk 1 above)
Comment (teisenbe) [reply MainUI.jsx]: Yeah, the property storage stuff is setup to share across all users using the same game UUID, so all windows viewing the same game would see the stats toggle if we wire it through storage.  I could go either way on it being persisted, but right now I believe we persist everything except the menu visibility to all instances.

**Comment by teisenbe** [comment]
I would be fine with landing what you have now and iterating, as long as it keeps defaulting to off.  @drakedevel: any issue?

**Comment by johngro** [comment]
k; FWIW, I messed around with this some tonight and have addressed many of the issues/requests.  In particular, I have...

- Made StatIcon into its own functional component in its own file.
- Added a LUT to StatIcon to help with some of the quirks (eg; to invert or not, and to substitute icon name or not)
- Handled the case of a numeric attribute with a modifier attached (eg; Retaliate 2: Range 3)
- Handled the case of pure text attributes (eg, Attackers gain disadvantage)
- Put boolean attributes on their own lines grouped by mob type.
- Rearranged the int stats to reflect what is on the actual in-game mob cards.  eg; three colums, normal numbers on the left, elite numbers on the right, icon in the middle, substitute '-' for 0.
- Made an attempt to organize my CSS a bit more.

I like the 3 column look now that it is in place.  The CSS still has trouble when laying out for a wide browser (instead of a tablet in portrait mode), but I'm hoping that I can land this and have you folks tweak it until it looks better.

The bool-stat-per-line seems fine, but the N vs. E letters feel a bit out of place.  I think I might want to mess around with...
- Putting one type of bool per line (like the ints are now)
- Putting icons in the left/right columns to show normal/elite support (again, like the ints are now)
- Showing a "-" to indicate a lack of support (again, like the ints).
- Leaving the center column blank for bools (the icon(s) will be on the left/right)

I don't want to do that now, however.  It is something I can play with after this version lands.

Next steps for me is to squash some changes, push to my branch, and update the pull request.  I'm going to wait until tomorrow evening to do that, however.  Right now, I'm gonna get some sleep.

**Comment by drakedevel** [comment]
Hello @johngro, and thanks for spending the time on this! I'm a big fan of the idea.

It's not currently automated (will take a look at getting some CI this evening) but there are linters configured for JS and Sass that you can run with `yarn run lint` and `yarn run lintsass`. If that's a pain (e.g. you're relying on the Docker build) don't worry about it, I can fix it up after I get CI running.

@teisenbe no objections to landing + iterating on polish.

**Comment by johngro** [comment]
ah, good to know.  I've been trying to be good about running eslint on all of the files that I have touched in order to keep in line with style and best practice.  It should be noted that my day job is writing device drivers and real-time audio systems in C++, so most of this is pretty new to me.  To be clear, my current flow after making a change and wanting to try it out locally has been just...

"npx eslint <files_ive_touched> && npx webpack"

If there is a better best practice, please do let me know.  As I said, I'm something of a foreigner in this world, so anything you can do to help me learn the customs and manners is much appreciated.

**Comment by drakedevel** [comment]
This world is kind of a disaster :wink: JS style all looks good to me from a quick overview. I did notice a vendor prefix in the stylesheets; the build process automatically inserts those, and I think the stylesheet linter should whine about it which is why I mentioned the linters :smiley: 

RE `eslint`: That invocation will definitely work for linting JS changes! I usually like to set up my editor to do this for me (I use Emacs, but there should be options for most editors), but running it manually works fine. There are some [scripts in package.json](https://github.com/drakedevel/gloomycompanion/blob/modernize/package.json#L9) (which can be run by `npm run <name>` or `yarn run <name>`), the `lint` one will automatically run `eslint` on all JS(X) files, which can save you from having to specify which files you've touched.

There's also a linter for stylesheets called `sass-lint`. The `lintsass` script will run it automatically on everything, or you can do run the relevant command manually. I also have Emacs set up to run this for me.

RE `webpack`: Running with no parameters should work fine, but is a slightly different configuration than I generally use (slightly slower build, less debug output, different source map). There are `build:dev` and `build:prod` scripts that pass the parameters/environment that I use for debug and release builds.

A far superior development experience can be had with the `serve` script. That will spawn a local webserver that will watch for source changes, automatically and incrementally re-build, and then **live update the code in your browser** (so-called Hot Module Reloading). For CSS changes this is quite reliable, for JS changes it's a little more hit-and-miss (sometimes has to give up and reload the tab, worst case you can do so manually). It really speeds up iteration for me.

**Comment by drakedevel** [comment]
@johngro I just landed a Travis CI config and fixed some outstanding ESLint errors; unfortunately it looks like the latter introduced a merge conflict. But, when you rebase, you should get a Travis build result!

**Comment by johngro** [comment]
ok, this should be ready for another look now.  I ran it through the linters; the JS should be clean.  The sass linter gave me some warnings, both about my stuff as well as some stuff I didn't touch.  I cleaned up the warnings it gave for the stuff I added, but left the other stuff as is for now.  If you want, I can take a swag at getting to 0 warnings and send a follow up CL.

FWIW - I want to clean up how I did the integer base stats parsing.  It got a little messy at the end there dealing with the special case of extra modifiers (retaliate 2: range 2), but I'll handle that in a follow up as well.  I'd rather get this in first so I can send out the fix for showing the flip animation when unhiding as well.

If it is easier, I can issue pull requests for sequences of CLs all at once.  This would be pretty typical for my day-to-day professional flow where we use gerrit for code review, but I'm unclear on what normal behavior on github would be when it comes to stacks of CLs.

**Comment by teisenbe** [comment]
Github makes it decent to look at the changes in each diff of a stack, I think.  I'd be fine with the whole stack in a single PR; the UI gives the person doing the final merge the ability to squash or rebase as appropriate

File: MainUI.jsx [hunk 4]
```diff
@@ -115,4 +149,8 @@ export default withStorage(MainUI, {
     path: 'mod_deck_hidden',
     deserialize: value => (typeof value === 'boolean' ? value : true),
   },
+  showBaseStats: {
+    path: 'mod_deck_hidden',
```
Comment (teisenbe) [inline MainUI.jsx:153]: the path should be changed to something like `show_base_stats`.  This value corresponds to part of the key in the Firebase db

File: AbilityDeck.jsx [hunk 5]
```diff
@@ -86,22 +130,143 @@ export default function AbilityDeck(props) {
...
-      className={props.hidden ? css.hiddenDeck : css.cardContainer}
-      onClick={props.onClick}
+    <div
+      className={props.hidden ? css.hiddenDeck : null}
     >
-      {topDraw ? renderCard(topDraw, -7, [css.draw], false) : null}
-      {topDiscard ? renderCard(topDiscard, -3, [css.discard, css.pull], true) : null}
-      {sndDiscard ? renderCard(sndDiscard, -4, [css.discard, css.lift], true) : null}
-    </ButtonDiv>
+      <div className={css.rowContainer}>
```
Comment (teisenbe) [inline AbilityDeck.jsx:256]: This breaks the unfortunately fragile `refresh_ui()` function in Tableau.jsx, which is what handles card text scaling for different size windows.  I'm not particularly opposed to the structure changes here, but we should update `refresh_ui()` to deal with it.  I'll open a separate issue to track the improvement of getting rid of `refresh_ui`.

The quick test for the resizing logic is to flip a card over and resize your window until the text and cards start shrinking, which should happen when the window is too narrow for a single ability deck

**Review by teisenbe — COMMENTED** [review]
The rest looks good to me; @drakedevel might want to take a look to make sure there aren't any non-idiomatic things/other hazards

File: AbilityDeck.jsx (hunk 5 above)
Comment (teisenbe) [reply AbilityDeck.jsx:256]: Oh, also I'd recommend rebasing before fixing that, since I just landed a fix for a different issue in refresh_ui()

Comment (johngro) [reply AbilityDeck.jsx:256]: k, I think that I tracked it down.  The problem wasn't really with refresh_ui, it was the fact that the container holding the card was not shrinking when the window became too narrow.  If you slap a max-width on it which says that it cannot be wider than the view port, the card can shrink again, the the font scaling hack seems to do its job again.

**Comment by johngro** [comment]
scaling should be better now.  I also added my fix for issue #4, as well as a CL with a small tweak to the gitignore.
